evolver/hardware/standard/pump.py

```diff
@@ -0,0 +1,112 @@
+from copy import copy
+
+from pydantic import Field
+
+from evolver.base import BaseConfig
+from evolver.hardware.interface import EffectorDriver, VialBaseModel
+from evolver.hardware.standard.base import SerialDeviceConfigBase
+from evolver.serial import SerialData
+
+
+class GenericPump(EffectorDriver):
+    """Pump control for evolver fluidics directly by pump ID.
+
+    This driver controls either or both of standard and IPP pumps according to the
+    arduino fluidics module. Pumps are indexed by pump ID (either position in array or
+    IPP pump number) and not vial, each must be controlled separately.
+
+    Specify the index of any IPP pumps via the config parameter ipp_pumps, this should
+    be an array of those pump IDs that are IPP. Please note that IPP pumps reserve 3
+    slots (1 for each solenoid), so if pump 0 is IPP, 1 and 2 cannot be assigned as
+    non-IPP pumps.
+    """
+
+    class Config(SerialDeviceConfigBase):
+        ipp_pumps: bool | list[int] = Field(False, description="False (no IPP), True (all IPP), or list of IPP ids")
+
+    class Input(BaseConfig):
+        pump_id: int
+        flow_rate: float
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.ipp_pumps is True:
+            self.ipp_pumps = list(range(self.slots / 3))
+        elif not self.ipp_pumps:
+            self.ipp_pumps = []
+
+    @property
+    def serial(self):
+        return self.serial_conn or self.evolver.serial
+
+    def set(self, input):
+        self.proposal[input.pump_id] = input
+
+    def commit(self):
+        inputs = copy(self.committed)
+        inputs.update(self.proposal.items())
+        cmd = [b"--"] * self.slots
+        for pump, input in inputs.items():
+            if pump in self.ipp_pumps:
+                # TODO: calibation transform here. The IPP pumps should return array of
+                # hz values per solenoid number - containing 3 of them.
+                hz_a = [input.flow_rate] * 3
+                for solenoid, hz in enumerate(hz_a):
+                    # solenoid is 1-indexed on hardware
+                    cmd[pump * 3 + solenoid] = f"{hz}|{pump}|{solenoid+1}"
+            elif any(pump - i < 3 for i in self.ipp_pumps):
+                raise ValueError(f"pump slot {pump} reserved for IPP pump, cannot address as standard")
+            else:
+                # TODO: calibration transform here. The time_to_pump is the target variable
+                # where pump_interval would presumably be what the pump_interval was set to
+                # during calibration and is fixed post-calibration.
```

Comment:
We need to discuss this for implications impacting #44. I don't think this will lean itself to a clean and extensible 
 calibration framework if we need to scatter the calibration in an ad-hoc manner depending on the device code.

Ideally, the calibration/transform needs to have happened before this gets called. However, the above might also add to the merit of separating the semantics of calibration from something like unit transformation.